substrate/frame/meta-tx/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Meta Tx
+//!
+//! TODO docs
+
+#![cfg_attr(not(feature = "std"), not_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub use pallet::*;
+
+use frame_support::{
+	dispatch::{DispatchInfo, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::OriginTrait,
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::traits::TransactionExtension;
+use sp_runtime::traits::TransactionExtensionBase;
+use sp_runtime::traits::{Dispatchable, IdentifyAccount, Verify};
+
+/// Meta transaction type.
+// TODO: The `MetaTx` type is similar to `sp_runtime::generic::UncheckedExtrinsic`. However,
+// `MetaTx` cannot replace generic::UncheckedExtrinsic because we need to box the call type,
+// given that `MetaTx` is used as an argument type for a call.
+#[derive(Encode, Decode, PartialEq, Eq, TypeInfo, Clone, RuntimeDebug)]
+pub struct MetaTx<Address, Signature, Call, TxExtension> {
+	proof: Proof<Address, Signature>,
+	call: Box<Call>,
+	tx_ext: TxExtension,
+}
+
+impl<Address, Signature, Call, TxExtension> MetaTx<Address, Signature, Call, TxExtension> {
+	pub fn new_signed(
+		address: Address,
+		signature: Signature,
+		call: Call,
+		tx_ext: TxExtension,
+	) -> Self {
+		Self { proof: Proof::Signed(address, signature), call: Box::new(call), tx_ext }
+	}
+}
+
+#[derive(Encode, Decode, PartialEq, Eq, TypeInfo, Clone, RuntimeDebug)]
+pub enum Proof<Address, Signature> {
+	Signed(Address, Signature),
+	// TODO `General` as in `sp_runtime::generic::Preamble`.
+}
+
+pub type MetaTxFor<T> = MetaTx<
+	<<T as Config>::PublicKey as IdentifyAccount>::AccountId,
+	<T as Config>::Signature,
+	<T as Config>::RuntimeCall,
+	<T as Config>::TxExtension,
+>;
+
+pub type SignedPayloadFor<T> =
+	sp_runtime::generic::SignedPayload<<T as Config>::RuntimeCall, <T as Config>::TxExtension>;
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The overarching call type.
+		type RuntimeCall: Parameter
+			+ GetDispatchInfo
+			+ Dispatchable<
+				Info = DispatchInfo,
+				PostInfo = PostDispatchInfo,
+				RuntimeOrigin = Self::RuntimeOrigin,
+			> + IsType<<Self as frame_system::Config>::RuntimeCall>;
+		/// Signature type for meta transactions.
+		type Signature: Parameter + Verify<Signer = Self::PublicKey>;
+		/// Public key type used for signature verification.
+		type PublicKey: IdentifyAccount<AccountId = Self::AccountId>;
```

Comment:
unrelated to this PR: Kind of a pitty that we can't fetch `PublicKey` and `Signature` from `AccountId`. The reverse is possible. You can always go from `Signature` (eg. `MultiSignature`) to `PublicKey` and then `AccountId` via `IdentifyAccount`.

Reply:
in this scope I do not even need a public key type, but have no other way to ensure the signature type implemented over the account id the system operates with.